runelite-client/src/main/java/net/runelite/client/plugins/inferno/InfernoOverlay.java

```diff
@@ -1,112 +1,480 @@
+				case 6:
+					color = new Color(0, 255, 255);
+					break;
+				default:
+					continue;
+			}
+
+			OverlayUtil.renderPolygon(graphics, tilePoly, color);
+		}
+	}
+
+	private void renderTicksOnNpc(Graphics2D graphics, InfernoNPC infernoNPC)
+	{
+		final Color color = (infernoNPC.getTicksTillNextAttack() == 1
+			|| (infernoNPC.getType() == InfernoNPC.Type.BLOB && infernoNPC.getTicksTillNextAttack() == 4))
+			? infernoNPC.getNextAttack().getCriticalColor() : infernoNPC.getNextAttack().getNormalColor();
+		final Point canvasPoint = infernoNPC.getNpc().getCanvasTextLocation(
+			graphics, String.valueOf(infernoNPC.getTicksTillNextAttack()), 0);
+		OverlayUtil.renderTextLocation(graphics, String.valueOf(infernoNPC.getTicksTillNextAttack()),
+			plugin.getTextSize(), plugin.getFontStyle().getFont(), color, canvasPoint, false, 0);
+	}
+
+	private void renderNpcLocation(Graphics2D graphics, InfernoNPC infernoNPC)
+	{
+		final LocalPoint localPoint = LocalPoint.fromWorld(client, infernoNPC.getNpc().getWorldLocation());
+
+		if (localPoint != null)
+		{
+			final Polygon tilePolygon = Perspective.getCanvasTilePoly(client, localPoint);
+
+			if (tilePolygon != null)
+			{
+				OverlayUtil.renderPolygon(graphics, tilePolygon, Color.BLUE);
+			}
 		}
-		return null;
 	}
 
-	// renders text location
-	private static void renderTextLocation(Graphics2D graphics, InfernoNPC actor, String text, Color color)
+	private void renderDescendingBoxes(Graphics2D graphics)
 	{
-		graphics.setFont(new Font("Arial", Font.BOLD, 15));
-		Point textLocation = actor.getNpc().getCanvasTextLocation(graphics, text, actor.textLocHeight + 40);
-		if (Strings.isNullOrEmpty(text))
+		for (Integer tick : plugin.getUpcomingAttacks().keySet())
 		{
-			return;
+			final HashMap<InfernoNPC.Attack, Integer> attackPriority = plugin.getUpcomingAttacks().get(tick);
+			int bestPriority = 999;
+			InfernoNPC.Attack bestAttack = null;
+
+			for (InfernoNPC.Attack currentAttack : attackPriority.keySet())
+			{
+				final int currentPriority = attackPriority.get(currentAttack);
+				if (currentPriority < bestPriority)
+				{
+					bestAttack = currentAttack;
+					bestPriority = currentPriority;
+				}
+			}
+
+			for (InfernoNPC.Attack currentAttack : attackPriority.keySet())
+			{
+				//TODO: Config values for these colors
+				final Color color = (tick == 1 && currentAttack == bestAttack) ? Color.RED : Color.ORANGE;
+				final Widget prayerWidget = client.getWidget(currentAttack.getPrayer().getWidgetInfo());
+
+				int baseX = (int) prayerWidget.getBounds().getX();
+				baseX += prayerWidget.getBounds().getWidth() / 2;
+				baseX -= BLOB_WIDTH / 2;
+
+				int baseY = (int) prayerWidget.getBounds().getY() - tick * TICK_PIXEL_SIZE - BLOB_HEIGHT;
+				baseY += TICK_PIXEL_SIZE - ((plugin.getLastTick() + 600 - System.currentTimeMillis()) / 600.0 * TICK_PIXEL_SIZE);
+
+				final Polygon blob = new Polygon(new int[]{0, BLOB_WIDTH, BLOB_WIDTH, 0}, new int[]{0, 0, BLOB_HEIGHT, BLOB_HEIGHT}, 4);
```

Comment:
this could probably be a static final field

Or even better, a rectangle. OverlayUtil isn't really needed for this, you can just use `Graphics.draw(Shape)`